### DISCUSSION TITLE
etcdClient / etcdServer compatibility during upgrade.

### DISCUSSION THREAD

**Discussion by jovajiv** [etcd-io/etcd#16305] [Q&A]
Hi All,

We're currently planning on upgrading to etcd 3.4 , our current status is  :
1. etcdClient - 3.3.13 ( edit : not etcdctl, just client side code using the following golang api https://github.com/etcd-io/etcd/blob/main/client/v3/client.go )
2. etcdServer cluster - 3.3.24 (edit:5 x pod Kubernetes cluster)

I've reviewed https://etcd.io/docs/v3.4/upgrades/upgrade_3_4/
but still have some open questions.

My question is, in which order should the upgrade occur ?
1. Do we upgrade the etcdServer first ? 
2. Is there any rush to upgrade the etcdClient once the server side is done ?   
3.do we have backwards compatibility to older client side versions (given that server upgrade already went through), allowing us to remain with 3.3.13 client side version for the time being ?
4. if you have any upgrade emphasis / relevant resources, please feel free to share.

Thank you.

**Comment by jmhbnz** [comment]
Hey @jovajiv - Thanks for your question. 

How many nodes are in your etcd cluster?  If more than one, yes you start with one server and ensure the cluster will become healthy using this new version member, refer step 3 and 4 here: https://etcd.io/docs/v3.4/upgrades/upgrade_3_4/#step-3-stop-one-existing-etcd-server 

When you say client, are you referring to `etcdctl`?  If so, please take note of the major change described here: https://etcd.io/docs/v3.4/upgrades/upgrade_3_4/#make-etcdctl_api3-etcdctl-default.   

Without a lot more context on the setup it's hard to provide any further advice. I would suggest taking some backups and trying out the upgrade procedure from end to end in a test environment.

**Reply by jovajiv** [reply]
hi @jmhbnz , Thanks for the quick response.

setup:
we have a 5 X pod etcd cluster on Kubernetes

when i say etcdClient, I'm referring to our usage of :
https://github.com/etcd-io/etcd/blob/main/client/v3/client.go
Which we use for client creation and client side code (Golang based).
